### ISSUE TITLE
All PACs Disappear When the Owner dies in Multiplayer

### ISSUE THREAD

**Issue by TheEnemyCube** [CapsAdmin/pac3#270]
My issue is as the title states, for some reason all PACs will disappear upon death of the owner. I know this is clientside as I heard my death sound in another person's mic while they were talking, but I could not see/hear my PAC and death sound when I died. This is incredibly frustrating and I have not been able to find a fix for it, so I hope to find it here.

EXTRA INFO:
Yes, I have gone into singleplayer and changed pac_sv_hide_outfit_on_death to 0, but it seems to reset to 1 when I join any server that has PAC3 enabled despite the server not having that option set as such.
My PAC does not disappear when I try to test this in singleplayer, so this is definitely something to do with multiplayer.
Another thing that I forgot to add in my sleepless state when I made this post was the fact that, in singleplayer, pac_sv_hide_outfit_on_death is already set to 0, so I'm not sure what it is about multiplayer that is doing this.

**Comment by LUModder** [comment]
I don't see why this happens. If pac_sv_hide_outfit_on_death is default 1 then why do most servers have it set to 1 and never have this issue? Also it's not going to change for you personally because it's a server cvar, that's what the sv in the name means.

**Comment by Hazbelll** [comment]
The default value of `pac_sv_hide_outfit_on_death` is 0. Two Clients on my Server both asked me why it was changed to 1, to which I had no idea, I've never touched it.

I checked the value of the cvar on the Server and on my own Client. Both were set to 0, as was everyone else. I changed the cvar value to 1 and then back to 0 on the Server and the value was replicated correctly to the two clients who had it set to 1.

I asked them to quit Garry's Mod, start it again and rejoin the Server, the value was still 0 for them, as it should be.

I'd thought I'd mention it here, considering the OP has brought up the cvar not replicating its value correctly on Servers which have it at its default value, although I can't say anything about their actual issue.

**Comment by TheEnemyCube** [comment]
I managed to get the issue fixed using the fix Hazbelll suggested above me, so I'm going to close the issue since there's nothing that needs to be fixed anymore.

**Comment by Hazbelll** [comment]
I also found out today it's PonyScape (a DarkRP server) causing it. If you join it, the cvar value gets messed up. I still have no clue how/why, but if you don't join that Server, you'll be fine.

**Comment by bgpony** [comment]
Hello! I'm Background. The server co-owner of Ponyscape, Gabriel, decided to use this in an attempt to increase performance for players.
I was unaware that this was an issue that would transition from server to server. Please forgive my rudimentary understanding of LUA and networking, but sv stands for server variable, a value for server operators to change, and cannot be changed by clients. Clients having this issue is not something I would take responsibility for, as it is the result of the value not being properly reset to 0, even if it is changed back to 0 on the server.
If it cannot be changed except by a server operator, I do not see its purpose otherwise. I don't see why servers shouldn't change server variables without the consent of people who willingly connect to the server if there is no malicious intent, when the issue that is created is due to an error in coding in the game. Regardless, the solution is very simple, and we will not be using this variable on the server because of this issue.

**Comment by LUModder** [comment]
The reason why it's so broken is because the flags aren't set up properly for the cvar. That's the only reason why it breaks so badly.

**Comment by Hazbelll** [comment]
The `FCVAR_SERVER_CAN_EXECUTE` flag gets automatically set with `FCVAR_REPLICATED`, but it wouldn't surprise me if that's not working correctly, considering that bug they had which got patched awhile ago (knowing GMod). Odd that it persists between servers though.

In this instance, said cvar is a server cvar, which the clients only replicate, so it's a legitimate mistake if it has such odd side-effects which are not of their control or fault.

However, It doesn't help that said server does alter cvars in the client realm without consent of the end-user (which have the `FCVAR_ARCHIEVE` flag, so get saved to disk) and mucking with some internal networked variables used by the game itself. That's asking for trouble.

**Comment by Moonakins** [comment]
This is a PAC3 issue, why are you bringing up things that have nothing to do with it, and talking shit about a server that you refuse to even talk to? You want to give us advice on how to do what we want without altering cvars, of which we do almost none of, then do it to our face. You complain about pony drama then you go and talk shit about us somewhere where it really doesn't need to be. This issue has had nothing to do with us altering PAC or us changing a client convar, it's just something that the client was not properly replicating.

**Comment by Hazbelll** [comment]
PAC's repo isn't the most appropriate place to be lashing out. As I originally stated, those who had this issue, when asked what servers they had played on recently, stated either yours or one other. I visited that one first, the issue didn't occur. I then visited yours, the issue did occur.

I tried to replicate the issue myself with a client, by setting the cvar from its non-default value, having them leave, setting it back to its default and seeing if their client would replicate it correctly. They did. Of course, that could've been a fluke. I did only test it once, but by this stage I figured you had just done something.

However, I was already aware you modify some clientside cvars without consent ("almost none" is not "absolute none"), but also more recently your internal networked variables had been tampered with. An sbox_maxprops limit of 0 for example, yet could still spawn props. The NWVar of the Usergroup for "superadmins" altered to be "admin". I don't know what else you could've been doing, but you *could* have easily been modifying the behaviour of PAC internally, which lead to undesired side-effects.

However as also said, a GMod update not so long ago fixed replicated cvars not replicating to clients correctly. This could still be the case, which as has also been said, and if so; is not your fault or responsibility.

The rest of your response is unrelated to this issue and not worth a response. I'll apologise if you felt like this was an "attack" however. That's understandable.

**Comment by Moonakins** [comment]
It's the pony community. Accusing servers of something they aren't even doing intentionally doesn't help.

Almost none is Fadmin_ShowChatNotifications, which should default to 0 or be a server controlled function anyway (Like ULX), no one likes their chatbox filled with things an admin is doing.
Didn't realize sbox_maxprops was 0, probably would have helped if you bothered to mention it to us.
Superadmins displaying as admin doesn't have anything to do with networked variables as far as I'm aware, they simply look the same in the fadmin scoreboard, but I could be wrong, I didn't do it.
Accusing us of something without talking to us at all about it is viewed as an attack by most people, I'd imagine.
I'll stop posting as I've gotten my views across, but thanks for making me aware of something as well as attacking our integrity without so much as talking to us, Hazbelll. You have the means.

**Comment by LUModder** [comment]
> I'll stop posting as I've gotten my views across, but thanks for making me aware of something as well as attacking our integrity without so much as talking to us

You're the one who lashed out but I don't want to start shit again.

**Comment by Hazbelll** [comment]
> It's the pony community.

No, it's just you here making a scene. I don't know how any ties with any community comes in here. This is an issue tracker totally unrelated to such.

> Accusing servers of something they aren't even doing intentionally doesn't help.

Already explained this above.

> which should default to 0 or be a server controlled function anyway (Like ULX), no one likes their chatbox filled with things an admin is doing.

Modifying the values of archived cvars is a very ugly way to do this, and frankly, very rude since you don't even provide the option. There's far more elegant ways to achieve what you want without doing so, whether force the value in-code (not touching the cvar) or providing the client an easy way to configure such, like through a menu.

This isn't the only time you've done it though. You did it to force enable the PlayX on clients in the past too. I'm unsure if anything else was.

> Didn't realize sbox_maxprops was 0, probably would have helped if you bothered to mention it to us.

Not a single player from your server reported this to you? Did you not check for bugs after editing whatever it is you edited in a test server or something? I find it impossible to believe you were unaware of the issue, as at least one of those must've happened. You shouldn't even need me to raise such an obvious error to you.

> Superadmins displaying as admin doesn't have anything to do with networked variables as far as I'm aware, they simply look the same in the fadmin scoreboard

Something has been altered somewhere, because "superadmin" does not show up as "admin" anywhere within stock Garry's Mod. The actual usergroup string is a networked variable also.

> Accusing us of something without talking to us at all about it is viewed as an attack by most people

Already explained this above aswell. No, it was not an attack, but a bad impression considering all that has been said above.

If you'd wish to discuss any of this further, for the sake of all those watching this repo, take it with me directly elsewhere, such as Steam (and don't bite me head off about it). If you genuinely want help, I'll be happy to provide it. Otherwise, that'll be all here.

**Comment by bgpony** [comment]
As I stated, the issue is resolved - We have a solution for it. As I feel, any client who connects to the server is consenting to the settings of the server. There is plenty of content on servers players would probably want to go without. e.g. your player controller, NSFW images that are downloaded in some maps. We do not use nor intend to do anything malicious on our server. If you feel we do bad practices, I'd rather you take that up with us than post even a sliver of that in a repo that has nothing to do with our server.

**Comment by Radray** [comment]
I looked into this issue and the issue is not with PS or PAC3.
Replicated convars created with lua has never worked as intended. I assume they recently tried to fix it since it would not network a non default convar to players joining after it was changed.

What it seems like now is that the problem has been reversed.
Setting the convar to a non-default value will set it for all connected and joining clients.
Setting the convar back to its default value, it will only set it for connected clients, NOT joining clients.

Setting FCVAR_SERVER_CAN_EXECUTE is not going to solve this issue.

The code below will let you define a default value in lua that does not cause this issue:

```
local cvar = CreateConVar("cvar_test", -1, {FCVAR_REPLICATED, FCVAR_ARCHIVE})
local default = 0
if (cvar:GetInt() == -1) then cvar:SetInt(default) end
```
